aspnetcore/mvc/views/razor.md

```diff
@@ -512,92 +585,279 @@ If "rick@contoso.com" is passed in the model, the view generates the following H
+
+For information on the properties and event handling behavior of the events in the preceding table, see [UIEventArgs](https://github.com/aspnet/AspNetCore/blob/master/src/Components/Components/src/UIEventArgs.cs) in the reference source.
+
+**Lambda expressions**
+
+Lambda expressions can also be used:
 
 ```cshtml
-@attribute [Authorize]
+<button @onclick="@(e => Console.WriteLine("Hello, world!"))">Say hello</button>
 ```
 
-### @namespace
-
-The `@namespace` directive sets the namespace of the class of the generated page or view:
+It's often convenient to close over additional values, such as when iterating over a set of elements. The following example creates three buttons, each of which calls `UpdateHeading` passing an event argument (`UIMouseEventArgs`) and its button number (`buttonNumber`) when selected in the UI:
 
 ```cshtml
-@namespace Your.Namespace.Here
+<h2>@message</h2>
+
+@for (var i = 1; i < 4; i++)
+{
+    var buttonNumber = i;
+
+    <button class="btn btn-primary"
+            @onclick="@(e => UpdateHeading(e, buttonNumber))">
+        Button #@i
+    </button>
+}
+
+@code {
+    private string message = "Select a button to learn its position.";
+
+    private void UpdateHeading(UIMouseEventArgs e, int buttonNumber)
+    {
+        message = $"You selected Button #{buttonNumber} at " +
+            $"mouse position: {e.ClientX} X {e.ClientY}.";
+    }
+}
 ```
 
-If a page or view imports API with an `@namespace` directive, the original file's namespace is set relative to that namespace. 
+> [!NOTE]
+> Do **not** use the loop variable (`i`) in a `for` loop directly in a lambda expression. Otherwise the same variable is used by all lambda expressions causing `i`'s value to be the same in all lambdas. Always capture its value in a local variable (`buttonNumber` in the preceding example) and then use it.
 
-If *MyApp/Pages/\_ViewImports.cshtml* contains `@namespace Hello.World`, the namespace of pages or views that import the `Hello.World` namespace is set as shown in the following table.
+### \@key
 
-| Page (or view)                     | Namespace               |
-| ---------------------------------- | ----------------------- |
-| *MyApp/Pages/Index.cshtml*         | `Hello.World`           |
-| *MyApp/Pages/MorePages/Bar.cshtml* | `Hello.World.MorePages` |
+The `@key` directive attribute causes the Razor components diffing algorithm to guarantee preservation of elements or components based on the key's value. For more information, see <xref:blazor/components#use-key-to-control-the-preservation-of-elements-and-components>.
 
-If multiple import files have the `@namespace` directive, the file closest to the page or view in the directory chain is used.
+### \@ref
 
-### @section
+Razor component references (`@ref`) provide a way to reference a component instance so that you can issue commands to that instance, such as `Show` or `Reset`. For more information, see <xref:blazor/components#capture-references-to-components>.
```

Comment:
Might be valuable to include the note form this section: https://docs.microsoft.com/en-us/aspnet/core/blazor/components?view=aspnetcore-3.0#capture-references-to-components